### PR TITLE
fix camera coord frames for all cameras and all robots for hw and sim

### LINKED ISSUE #669: review camera coordinate frames
there's a mismatch between `_link` and `_optical_frame` for the cameras we're using in cob_bringup
- realsense: OK, compliant with REP103 http://www.ros.org/reps/rep-0103.html (done in https://github.com/ipa320/cob_robots/pull/614)
- Asus/kinect: Not OK
- Sick Visionary: Not Ok
- usb head cam: to be verified

### PR DESCRIPTION
**Pull request by fmessmer**
fixes #669
fixes ipa320/cob4#553
requires https://github.com/ipa320/cob_common/pull/233

includes #715

@ipa-nhg @cagbal @ipa-fmw @ipa-rmb FYI

### CODE REVIEW

**Comment by fmessmer** [comment]
in https://github.com/ipa320/cob_robots/pull/720/commits/51439d99d1a3d45037946eb73376a393c393395c I added a `nodelet_manager` for simulation!
this was necessary as those cameras use the `nodelet`-variant of image_flip in the respective components-launch file...however, in simulation there was no nodelet_manager running so far...as this was part of the hw-driver...now we have all the `X_upright/X` topics in simulation, too

**Comment by fmessmer** [comment]
ToDo:
- headcam coordinate system for cob4-12 needs to be updated, too
  (i.e. final comparison of cob_robots bs. unity_robots)
- verify sick_3dcs and usb_cam topics and links with Hardware, e.g. tlabs

**Review by floweisshardt — DISMISSED** [review]
tested on cob4-12: OK

**Comment by fmessmer** [comment]
I consider this final now
I've tested this on hardware (`cob4-7` and `cob4-10`) as well as in simulation

Here are some screenshots of the live data from the robots:
**cob4-7**
head_cam
head_cam_live_data

sensorring (asus)
sensorring_live_data

torso_down (r200)
torso_down_live_data

torso_left (zr300)
torso_left_live_data

torso_right (zr300)
torso_right_live_data

**cob4-10**
head_cam
head_cam_live_data

sensorring_front (sick_3dcs)
sick_3dcs_live_data

sensorring_back (asus)
sensorring_back_live_data

torso_down (r200)
torso_down_live_data

torso_left (asus)
torso_left_live_data

torso_right (asus)
torso_right_live_data

**Comment by fmessmer** [comment]
I've opened an issue for the wrong transformations in the sick_3dcs PointCloud here: https://github.com/SICKAG/sick_visionary_t/issues/19

File: cob_hardware_config/robots/cob4-10/config/image_flip_params_sensorring_cam3d_front.yaml [hunk 1]
```diff
@@ -12,7 +12,7 @@ reference_frame: base_link
 
 # flip color image
 # bool
-flip_color_image: true
+flip_color_image: false
```
Comment (floweisshardt) [inline cob_hardware_config/robots/cob4-10/config/image_flip_params_sensorring_cam3d_front.yaml:15]: why false?

File: cob_bringup/components/cam3d_visionary_t.launch [hunk 2]
```diff
@@ -13,5 +14,13 @@
 		<arg name="yaml_name" value="$(arg name)"/>
 		<arg name="sim" value="$(arg sim)"/>
 	</include>
-
+	<include if="$(arg flip)" file="$(find cob_bringup)/drivers/image_flip.launch">
+		<arg name="robot" value="$(arg robot)"/>
+		<arg name="camera_name" value="$(arg name)"/>
+		<arg name="pointcloud_in" value="/$(arg name)/depth/points"/>
+		<arg name="pointcloud_out" value="/$(arg name)_upright/depth/points"/>
+		<!-- Sick3DCS does not have color image, and image flip cannot process images that are not CV_8UC3 -->
```
Comment (fmessmer) [inline cob_bringup/components/cam3d_visionary_t.launch:22]: @ipa-rmb FYI
I don't know whether this is intended...

**Review by floweisshardt — APPROVED** [review]
except for the ```flip_color_image: false``` for cob4-10, I'm OK with this PR.

File: cob_hardware_config/robots/cob4-10/config/image_flip_params_sensorring_cam3d_front.yaml (hunk 1 above)
Comment (fmessmer) [reply cob_hardware_config/robots/cob4-10/config/image_flip_params_sensorring_cam3d_front.yaml:15]: sick_3dcs does not have a color_image...@ipa-rmb @ipa-josh 
and image_flip complained about not being able to flip the depth/image topic (see https://github.com/ipa320/cob_robots/pull/720#discussion_r140308555)

**Comment by fmessmer** [comment]
@ipa-rmb @ipa-fmw
why do we actually need to flip the point cloud? it looks correct in rviz anyway...
does any node processing the point cloud depend on it being upright?
if this is not the case, we might safe some computational resources if we do not flip the point cloud, right?

**Comment by ipa-rmb** [comment]
Yes, because the pixel matrix should be the same for the color image and the point cloud. If I move pixels on the color image to a different location, then I would expect to find the corresponding XYZ pixel at the same pixel coordiates that I used in the color image. That is why the point cloud needs to be re-arranged as well.

File: cob_bringup/components/cam3d_visionary_t.launch (hunk 2 above)
Comment (ipa-rmb) [reply cob_bringup/components/cam3d_visionary_t.launch:22]: I do not care about that camera at the moment. Image flip would need to be extended to handle other types of images as well, the current code fpr 8UC3 images is speed optimized and not general. We can later add a slower general fall back solution for other image types if anyone needs that at all.

**Comment by ipa-rmb** [comment]
Transforms and images for cob4-10 look fine.

**Review by ipa-rmb — APPROVED** [review]

File: cob_hardware_config/robots/cob4-10/config/image_flip_params_sensorring_cam3d_front.yaml (hunk 1 above)
Comment (floweisshardt) [reply cob_hardware_config/robots/cob4-10/config/image_flip_params_sensorring_cam3d_front.yaml:15]: ok fine

Comment (ipa-rmb) [reply cob_hardware_config/robots/cob4-10/config/image_flip_params_sensorring_cam3d_front.yaml:15]: also fine with us, see comment above.

**Comment by ipa-rmb** [comment]
Can/shall I merge?

File: cob_bringup/components/cam3d_visionary_t.launch (hunk 2 above)
Comment (fmessmer) [reply cob_bringup/components/cam3d_visionary_t.launch:22]: I created a new issue for this: https://github.com/ipa320/cob_perception_common/issues/85

**Comment by fmessmer** [comment]
As I opened issues for all remaining issues, I'll merge all related PRs for now...
...in case anything is not as expected with this version, it needs to be investigated in a separate round...

**Pull request merged**